feat(swagger-editor): customizable codegen URLs

antoineauger opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
This PR aims at allowing users to customize the URLs used by the Swagger Editor codegen features (e.g., Generate Server and Generate Client).

The URLs that can now be customized are:
- swagger2GeneratorUrl: (https://generator.swagger.io/api/swagger.json by default)
- oas3GeneratorUrl: (https://generator3.swagger.io/openapi.json by default)
 - swagger2ConverterUrl: (https://converter.swagger.io/api/convert by default)

For Docker users, this is made possible by the introduction of 3 new environment variables: `URL_SWAGGER2_GENERATOR` `URL_OAS3_GENERATOR` and `URL_SWAGGER2_CONVERTER`, respectively.

For users who create a `SwaggerEditor` object, they can now provide an additional dict object to override these URLs (see `index.html` for an example)
```js
function SwaggerEditor(options, swaggerEditorConfigUrls = null) {
  // ...
  return SwaggerUI(mergedOptions)
}
```

### Motivation and Context

This PR can be interesting for:
- users already have their own local Swagger generator server (https://github.com/swagger-api/swagger-codegen#public-pre-built-docker-images);
- users don't want to enable the codegen features (for instance for security reasons).

Fixes #1488 #2127

/cc @fgreinacher @dlouzan

### How Has This Been Tested?

For Docker env variables, created several docker containers with various values for env variables `URL_SWAGGER2_GENERATOR`, `URL_OAS3_GENERATOR` and `URL_SWAGGER2_CONVERTER`. Then logging on startup:
```js
console.log(mergedOptions)
```

Also tested that setting these variables to `null` was disabling the Swagger Editor codegen features:
```
docker run -e URL_SWAGGER2_GENERATOR="null" -e URL_OAS3_GENERATOR="null" -e URL_SWAGGER2_CONVERTER="null" -d -p 9999:8080 swagger-editor
```


## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

### My PR contains... 
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [ ] Improvements (misc. changes to existing features)
- [x] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [ ] My changes do not require a change to the project documentation.
- [x] My changes require a change to the project documentation.
- [x] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [x] My changes can not or do not need to be tested.
- [ ] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [ ] All new and existing tests passed.

char0n commented:
@antoineauger thanks a lot for contributing. We'll look into the PR ASAP.

char0n commented on src/index.js:
```diff
@@ -65,9 +65,15 @@ const defaults = {
 }
 
-export default function SwaggerEditor(options) {
+export default function SwaggerEditor(options, swaggerEditorConfigUrls = null) {
```
What would be the reason for extending the `SwaggerEditor` function with another parameter called `swaggerEditorConfigUrls`?

`SwaggerEditor` function already supports following configuration options via `options` parameter ([see code here](https://github.com/swagger-api/swagger-editor/blob/2581d92dcf984bea0b6c19d96894f5f24dd1f03a/src/index.js#L63)):

```js
  swagger2GeneratorUrl: "https://generator.swagger.io/api/swagger.json",
  oas3GeneratorUrl: "https://generator3.swagger.io/openapi.json",
  swagger2ConverterUrl: "https://converter.swagger.io/api/convert",
```

So IMHO in order to support new ENV variables we don't need to make any changes to the code.

antoineauger replied:
Also updated the PR description.

char0n commented on index.html:
```diff
@@ -57,6 +57,11 @@
       ],
       queryConfigEnabled: false,
+    },
+    {
```
Correct me if I'm wrong, but this looks line unnecessary change. The configuration options are already present and have their default values set:

https://github.com/swagger-api/swagger-editor/blob/2581d92dcf984bea0b6c19d96894f5f24dd1f03a/src/index.js#L63

antoineauger replied:
You're right, see my answer: https://github.com/swagger-api/swagger-editor/pull/2910#discussion_r821852314

char0n commented on README.md:
```diff
@@ -129,6 +129,16 @@ docker run -d -p 80:8080 -v $(pwd):/tmp -e SWAGGER_FILE=/tmp/swagger.json swagge
+You can also customize the different endpoints used by the Swagger Editor with the following environment variables. For instance, this can be useful if you have your own Swagger generator server:
+
+Environment variable | Default value
+--- | --- | ---
```
This unfortunately doesn't generate the table. There are 3 columns in table header, but only 2 columns in table body.

char0n commented on README.md:
```diff
@@ -129,6 +129,16 @@ docker run -d -p 80:8080 -v $(pwd):/tmp -e SWAGGER_FILE=/tmp/swagger.json swagge
+`URL_OAS3_GENERATOR` | https://generator3.swagger.io/openapi.json
+`URL_SWAGGER2_CONVERTER` | https://converter.swagger.io/api/convert
+
+If you simply want to run the Swagger Editor locally without the Codegen features (Generate Server and Generate Client) you can simply set the above environment variables to `null`.
```
```suggestion
If you want to run the Swagger Editor locally without the Codegen features (Generate Server and Generate Client) you can set the above environment variables to `null` (`URL_SWAGGER2_CONVERTER=null`).
```

char0n commented on docker-run.sh:
```diff
@@ -40,6 +40,28 @@ if [[ "${QUERY_CONFIG_ENABLED}" = "true" ]]; then
 
+if [[ "${URL_SWAGGER2_GENERATOR}" ]]; then
+  if [[ "$URL_SWAGGER2_GENERATOR" != "null" ]]; then
+    sed -i "s|swagger2GeneratorUrl: 'https://generator.swagger.io/api/swagger.json'|swagger2GeneratorUrl: '${URL_SWAGGER2_GENERATOR}'|g" $INDEX_FILE
```
To add the `swagger2GeneratorUrl` as option to `SwaggerEditorBundle` function, following shell code needs to be introduced:

```suggestion
    sed -i "s|SwaggerEditorBundle({|SwaggerEditorBundle({\n      swagger2GeneratorUrl: '${URL_SWAGGER2_GENERATOR}',|g" index.html
```

This independently injects new options to `SwaggerEditorBundle` function and should be used through out the changes in `docker-run.sh` that this PR introduced.

antoineauger replied:
You're right, see my answer: https://github.com/swagger-api/swagger-editor/pull/2910#discussion_r821852314

char0n reviewed (changes requested):
Thanks for contributing! Please have a look at my code review comments.

antoineauger commented:
> Thanks for contributing! Please have a look at my code review comments.

Thanks for the review @char0n, this PR is way much simpler now.

I think I addressed all your comments, could you have another look at it? 🙏🏻

char0n reviewed (approved):
Nice work! Changes look good! I'll have to some local testing with changes from this PR before the merge.

char0n commented:
Testing went fine. All works as expected. THanks again!
